src/blas/impl/KokkosBlas3_gemm_impl.hpp

```diff
@@ -392,7 +399,7 @@ KOKKOS_INLINE_FUNCTION
 void impl_team_gemm_block(const TeamHandle& team, const ViewTypeC& C, const ViewTypeA& A, const ViewTypeB& B) {
   typedef typename ViewTypeC::non_const_value_type ScalarC;
 // GNU COMPILER BUG WORKAROUND
-#if defined(KOKKOS_COMPILER_GNU) || !defined(__CUDA_ARCH__)
+#if defined(KOKKOS_COMPILER_GNU) && (!defined(__CUDA_ARCH__) || !defined(__HIP_DEVICE_COMPILE__))
```

Comment:
Is this correct, it seems that you might want to replace `&&` should be replaced with `||`

Reply:
I am totally fine if the new behavior is what we want to implement, I just wanted to point out that the logic probably changed here

Reply:
@lucbv I noticed this was different than most of the other GCC bug workarounds involving const, which used ``defined(KOKKOS_COMPILER_GNU) && !defined(__CUDA_ARCH__)``. ``||`` would mean "anything that's not CUDA device code" which I don't think is what was intended.

Reply:
Yeah that makes sense to me, I just wanted to confirm as this was clearly changing the logic.
I'm good with this PR.